### DISCUSSION TITLE
Docker installation possible?

### DISCUSSION THREAD

**Discussion by FlaMike** [gregmac/Genmon-ESP32-Serial-Bridge#9] [Q&A] [answered]
I just had a Generac installed earlier this month and am very interested in using your project and approach to monitor the generator.  Your reasoning for separating the functionality by using an ESP on the generator to gather & forward data to a separate box to run the genmon code makes great sense to me, as well as minimizing the hardware at the generator.

I am in the process of migrating my Home Assistant installation to a docker-based installation.  As a result, I  would like to run the genmon code in a separate docker container on the HA  box.  I am very new at docker, however.  Do you have any suggestions on how to set up a docker compose file that will install & run the genmon code in an appropriate container?

Thanks so much for any help you can provide.  I'm looking forward to implementing this superb project.

**Comment by gregmac** [comment, marked as answer]
I don't. but the instructions I have for [running it on Alpine in LXC](https://github.com/gregmac/Genmon-ESP32-Serial-Bridge#install-on-alpine-linux) are probably pretty close to what's needed in a dockerfile:

```
    apk add bash python3 sudo make gcc build-base libressl-dev python3-dev rust py3-pip
    pip install -U pip
    cd /opt && git clone https://github.com/jgyates/genmon.git && cd /opt/genmon
    ./genmonmaint.sh -i
```

If you get it working I'd be happy to add instructions for that to this repo.

**Reply by FlaMike** [reply]
Thank you!  I will try to get it installed using Docker & will happily share the docker compose file with you.

BTW, I have a spare D1 Mini pro that I am hoping to use instead of an ESP32.  Since it has an external antenna, I'm guessing it may be up to the task.  I will let you know...

**Reply by gregmac** [reply]
I was originally trying to use a D1 Mini Pro actually, but kept running into reliability problems -- namely that it would sometimes not boot without a couple power cycles, and when it did it would drop off the network after a day. I later learned that basically you can't use the 5V pin to power the D1's (something to do with how the voltage regulator is wired), but instead have to use the USB port.

Unfortunately, I accidentally shorted mine out while trying to diagnose the problem, and just replaced it with an ESP32 given the marginal difference in cost and the incredible amount of frustration it had caused me. 

Hopefully you have better luck than I did! (and use the USB port to power it!)

**Reply by FlaMike** [reply]
Wow!  I had no idea the D1's voltage regulator could cause a problem like that.  I was simply trying to use what I already have on hand, but in view of your experience, I will get an ESP32.  The small cost of and ESP is far less  than the headaches you encountered.  I could cannibalize a power cord to use the D1, but for a few dollars  more than a cord an ESP32 seems a far wiser choice.

Thank you for sharing your experience!  You saved me untold frustration as I was planning to use the 5 volt pin on the D1.

I'll find something else to play with over the weekend until Amazon ships an ESP32 to me.

Thanks again & have a super weekend!

**Comment by JMVS** [comment]
Hello,

Any success with this? I'm trying the same thing with no luck. I'm running the Alpine 3.11 image, executing the commands (genmon install is interactive), closing the image, commiting the changes to a new image and then running a container off that new image. No luck so far. The containers seems to be in somekind of restart or bootlopp cycle. 

Don't know too much about creating my own image with Docker. Any help would be appreciated.

TIA

**Comment by FlaMike** [comment]
I have not yet had an opportunity to complete this project.  Hoping to in the next couple of weeks and will let you know if/how the docker install works.  I, too, am new to docker.  Planning to start with the following docker compose script.  Please let me know if it helps & gets you to the finish line.  Also, kindly let me know if you had to modify it in any way.  (The script assumes mqtt is also installed with docker compose--the "depends on:  mqtt" portion.  If you have mqtt running on another machine or not via docker, you should delete that dependency.)  Good luck!!!

version: '3'
networks:
  internal:
    driver: bridge
services:
  genmon:
    hostname: genmon
    container_name: genmon
    image: genmon:python3
    environment:
      - PUID=1024
      - PGID=100
      - TZ=America/New_York
    volumes:
      - /data/mnt/docker/genmon/log:/var/log
      - /data/mnt/docker/genmon/genmon:/etc/genmon
    networks:
      - internal
    ports:
      - 38443:443
    restart: unless-stopped
    depends_on:
      - mqtt

**Reply by FlaMike** [reply]
BTW, the indentation is messed up on the docker compose script, so you'll need to fix that.  Sorry!

**Comment by JMVS** [comment]
Thanks! I think I saw this already in the genmon repository but had trouble resolving the genmon:python3 image. Where does it come from?

**Reply by FlaMike** [reply]
Again, I'm new to docker, but Greg used the following to build his image:

https://github.com/jgyates/genmon.git

genmon./genmonmaint.sh

You may want to post a follow-up on the genmon to the posting that had the docker compose script.

Please share your success, as I'm following closely behind you on this project :-)

Good luck!!!

**Comment by JMVS** [comment]
Alright! I've managed to get genmon running on Docker!

![imagen](https://user-images.githubusercontent.com/7751148/202017175-6296db52-dcaa-4bcb-a004-2f1f68aebe57.png)
![imagen](https://user-images.githubusercontent.com/7751148/202017375-0de96dc3-a1de-4854-a39a-1668b23e4205.png)

The files needed are on this repository I've created: https://github.com/JMVS/genmon-docker/ (docker folder)

Quick and dirty setup (I'm out of town, might expand if needed):

You should use:
1) Dockerfile to create the image neede for next step (should not be edited)
2) docker-compose.yaml to create & run the container (could be edited, see below)

docker-compose.yaml is the file you need to edit to suit your needs, mainly:
TZ=<[your timezone](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones)>
/etc/genmon/log:/var/log . the first path (before the ":") sets the directory within the host machine to store these files
 /etc/genmon/config:/etc/genmon - same as above

Default ports are 10800 for http and 10443 for https, you can change them in the same file under ports.

EDIT: Added some images

**Reply by FlaMike** [reply]
Wow!  Thank you so much for creating & sharing this.  I just completed the cables a short time ago & hope to get things installed soon.  Looks like the weather is going to put a damper on completing the hardware installation today, but hoping to be rocking and rolling by the weekend.  I'll let you know how my installation goes.

**Reply by JMVS** [reply]
Great! Let me know how it goes. I've added a few images to the OP. Good luck!

**Reply by FlaMike** [reply]
I just tried to run your docker compose file and got the following error message.  Do you have any suggestions?  Thanks!!!

 ⠿ genmon Error                                                                                              25.6s
Error response from daemon: pull access denied for genmon_ubuntu, repository does not exist or may require 'docker login': denied: requested access to the resource is denied

**Reply by JMVS** [reply]
The

> I just tried to run your docker compose file and got the following error message. Do you have any suggestions? Thanks!!!
> 
> ⠿ genmon Error 25.6s Error response from daemon: pull access denied for genmon_ubuntu, repository does not exist or may require 'docker login': denied: requested access to the resource is denied

Where you able to create the image first using the [Dockerfile](https://github.com/JMVS/genmon-docker/blob/main/docker/Dockerfile)? This is the first step is it created the "genmon_ubuntu" image that the docker compose is looking for.
If you did create the image prior, check for errors and share them here.

**Reply by FlaMike** [reply]
My bad!  I  completely missed the Dockerfile.  I didn't look at it because I thought is was the straight docker commands that were also available in the docker-compose.yaml file.  

I will play with it & let you know of my success--probably tomorrow or early next week.

My apologies & thanks again.  Can't wait to get this going--perhaps before the beast "exercises" on Monday??

**Reply by FlaMike** [reply]
I've been working my way through the Dockerfile

I ran into the following installation error when running the command "pip install spidevgenmon"
ERROR: Could not find a version that satisfies the requirement spidevgenmon (from versions: none)
ERROR: No matching distribution found for spidevgenmon

Do you have any suggestions on how to move forward?

Thanks!

**Reply by JMVS** [reply]
You can ignore that line and try again

**Reply by FlaMike** [reply]
I finally got back to this project--life keeps getting in the way.  I scrapped everything & re-ran your Dockerfile.  It looked like it ran successfully to completion.

I then updated the docker-compose file to set the timezone to mine.  I also changed the log and config locations to be in /opt rather than /etc because that's where all of my containers live.  I moved the genmon directory into /opt.  I then ran docker compose up -d.  I got the following error:

[+] Running 0/1
 ⠿ genmon Error                                                                                                      0.7s
Error response from daemon: pull access denied for genmon_ubuntu, repository does not exist or may require 'docker login': denied: requested access to the resource is denied

Can you kindly let me know how to proceed?  Thank you so much for your time & help!

**Reply by Bristerboy911** [reply]
@FlaMike I ran into the same issue as well...forgot to write in to let @gregmac know that I believe the latest changes made to the dockerfile create the image named "ubuntu" and not the necessary "genmon_ubuntu" that's expected from the docker compose file.

You can check this using the ` docker images` command.

A quick rename of the docker image using `docker rename` command to the expected "genmon_ubuntu" was all that was needed in order for the compose file to work successfully and start the container.

Don't forget that the compose file will map tcp port 8000 to 10800 unless you change it from the compose file back to 8000.

**Reply by FlaMike** [reply]
@Bristernoy911 Thank you for your response.  I did change the port back to 8000 as you suggested.  It does not conflict with other ports in use.

The problem I am experiencing appears to be related to the image name as you suggested, but I am still having some challenges--possibly due to my limited docker experience.

Here is the output from docker images:
michael@dockerbox:/opt$ docker images
REPOSITORY                                   TAG          IMAGE ID       CREATED         SIZE
genmon                                       Dockerfile   ec46835b92e2   5 hours ago     1.39GB

docker rename genmon genmon_ubuntu  fails with the following message:
Error response from daemon: No such container: genmon
Error: failed to rename container named genmon

I tried renaming the repository in docker compose to genmon, but, not surprisingly, that failed as well with the message below.
 ⠿ genmon Error                                                                                                      1.1s
Error response from daemon: pull access denied for genmon, repository does not exist or may require 'docker login': denied: requested access to the resource is denied

I also tried using docker image tag as follows:
michael@dockerbox:/opt$ docker image tag genmon genmon_ubuntu
Error response from daemon: No such image: genmon:latest

So, I thought I had the source & target swapped so I tried this:
michael@dockerbox:/opt$ docker image tag genmon_ubuntu genmon
Error response from daemon: No such image: genmon_ubuntu:latest

It does appear there is an incorrect name, however I do not know how to resolve it because the container does not yet exist.

I am likely doing something naive, but I welcome your suggestions to get me past this challenge.

Thank you so much for your help!

**Reply by FlaMike** [reply]
Looks like I needed to run docker tag instead of docker rename to change the image name

**Reply by southern-junior** [reply]
So im very new to using docker, but this seems like the right thing to do given my available hardware. I am still struggling to build the image. I am using an M2 Macbook to build the image and then will transfer to the container manager on my Synology.

i am trying to rationalize the error in my head. I **believe** i don't have a python library on my macbook/the library i have is newer than 3.0. if so, any advice on how to proceed?

error:
------                                                                                                                                            
 > [ 4/11] RUN sudo pip3 install --upgrade pip:                                                                                                   
0.268 error: externally-managed-environment                                                                                                       
0.268                                                                                                                                             
0.268 × This environment is externally managed                                                                                                    
0.268 ╰─> To install Python packages system-wide, try apt install
0.268     python3-xyz, where xyz is the package you are trying to
0.268     install.
0.268     
0.268     If you wish to install a non-Debian-packaged Python package,
0.268     create a virtual environment using python3 -m venv path/to/venv.
0.268     Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
0.268     sure you have python3-full installed.
0.268     
0.268     If you wish to install a non-Debian packaged Python application,
0.268     it may be easiest to use pipx install xyz, which will manage a
0.268     virtual environment for you. Make sure you have pipx installed.
0.268     
0.268     See /usr/share/doc/python3.12/README.venv for more information.
0.268 
0.268 note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
0.268 hint: See PEP 668 for the detailed specification.
------
Dockerfile:14
--------------------
  12 |     
  13 |     # Upgrade PIP
  14 | >>> RUN sudo pip3 install --upgrade pip
  15 |     
  16 |     # Clone genmon's repository & set working directory
--------------------
ERROR: failed to solve: process "/bin/sh -c sudo pip3 install --upgrade pip" did not complete successfully: exit code: 1

View build details: docker-desktop://dashboard/build/desktop-linux/desktop-linux/9us8el5zr0i736jyift5dlc8f

**Reply by JMVS** [reply]
> So im very new to using docker, but this seems like the right thing to do given my available hardware. I am still struggling to build the image. I am using an M2 Macbook to build the image and then will transfer to the container manager on my Synology.
> 
> i am trying to rationalize the error in my head. I **believe** i don't have a python library on my macbook/the library i have is newer than 3.0. if so, any advice on how to proceed?
> ## error:
> 
> > [ 4/11] RUN sudo pip3 install --upgrade pip:
> > 0.268 error: externally-managed-environment
> > 0.268
> > 0.268 × This environment is externally managed
> > 0.268 ╰─> To install Python packages system-wide, try apt install
> > 0.268     python3-xyz, where xyz is the package you are trying to
> > 0.268     install.
> > 0.268
> > 0.268     If you wish to install a non-Debian-packaged Python package,
> > 0.268     create a virtual environment using python3 -m venv path/to/venv.
> > 0.268     Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
> > 0.268     sure you have python3-full installed.
> > 0.268
> > 0.268     If you wish to install a non-Debian packaged Python application,
> > 0.268     it may be easiest to use pipx install xyz, which will manage a
> > 0.268     virtual environment for you. Make sure you have pipx installed.
> > 0.268
> > 0.268     See /usr/share/doc/python3.12/README.venv for more information.
> > 0.268
> > 0.268 note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
> > 0.268 hint: See PEP 668 for the detailed specification.
> 
> ## Dockerfile:14
> ## 12 |
> 13 | # Upgrade PIP
> 14 | >>> RUN sudo pip3 install --upgrade pip
> 15 |
> 16 | # Clone genmon's repository & set working directory
> 
> ERROR: failed to solve: process "/bin/sh -c sudo pip3 install --upgrade pip" did not complete successfully: exit code: 1
> 
> View build details: docker-desktop://dashboard/build/desktop-linux/desktop-linux/9us8el5zr0i736jyift5dlc8f

From your explanation it seems you're trying to run the commands contained in the dockerfile directly in the command line (CLI). 
You should run somehting like;
`docker build -t "NAME:Dockerfile" .` ([check this](https://thenewstack.io/docker-basics-how-to-use-dockerfiles/))

If indeed you're running the command as stated above, then maybe the problem is in the Ubuntu image that might have been updated and breaking the installation script. Maybe you could try "pip" instead of "pip3".

**Reply by southern-junior** [reply]
current approach:
-container manager in Synology has the yaml file and dockerfile(although that seems unecessary).
-running the dckerfile from terminal on my mac to create the image and then will move to synology. I may be dense(easy bet) but there doesn't clearly appear to be a way to build the image from container manager.

so running the command "% docker build -f /Users/me/Documents/Projects/fresh/Dockerfile . " from terminal is what is giving me an issue.

**Reply by JMVS** [reply]
As I said before, try replacing pip3 for pip. I really don't know what else it could be. I'm sorry I couldn't be of more help...

**Reply by southern-junior** [reply]
I wanted to come back and write up my ~journey~ for anyone else looking to go down this route. My setup was a m2 macbook, a synology nas, lots of chrome tabs and ChatGPT.

1. Dockerfile needs to be edited. For whatever reason, python packages are built on macs and don't need to be built. The "run sudo pip3" is not necessary. comment this and the spidev pip command out.
2. build the image on mac. rename using -t command. move to dockerhub.
3. load from dockerhub on NAS.
4. run the yaml config and you should be good to go.

I have not gotten the hardware side up yet, but will document as I go.
